frontend/src/main/scala/bloop/dap/DebugServer.scala

```diff
@@ -0,0 +1,79 @@
+package bloop.dap
+
+import java.net.{ServerSocket, URI}
+
+import bloop.io.ServerHandle
+import bloop.logging.Logger
+import monix.eval.Task
+import monix.execution.{Cancelable, Scheduler}
+
+import scala.collection.mutable
+import scala.concurrent.Promise
+import monix.execution.cancelables.MultiAssignmentCancelable
+import monix.execution.atomic.AtomicBoolean
+
+final class StartedDebugServer(
+    val address: Task[Option[URI]],
+    val listen: Task[Unit]
+)
+
+object DebugServer {
+  def start(
+      runner: DebuggeeRunner,
+      logger: Logger,
+      ioScheduler: Scheduler
+  ): StartedDebugServer = {
+    /*
+     * Set backlog to 1 to recommend the OS to process one connection at a time,
+     * which can happen when a restart is request and the client immediately
+     * connects without waiting for the other session to finish.
+     */
+    val handle = ServerHandle.Tcp(backlog = 1)
+
+    val closedServer = AtomicBoolean(false)
+    val listeningPromise = Promise[Option[URI]]()
+    val ongoingSession = MultiAssignmentCancelable()
+
+    def listen(serverSocket: ServerSocket): Task[Unit] = {
```

Comment:
> Could you include an error? Does it contain a JDI output?
  
Didn't attach an error because it's just a timeout. Can you not reproduce it?
  
> Also, could you point me to those hanging issues? In the end, I might miss some.
  
I'll have another look in the second review round, but the most glaring ones were in `DebugSession` and `DebugServer`. Let's make sure that in these two abstractions, we have a clear distinction between: closing resources when everything exits successfully, cancelling and starting the logic, and that none of these race with each other.

Reply:
In the end I was not able to reproduce it on this particular test case. Let's hope the fixes I introduced  help.